Add GPU acceleration for morphology filters

nullhook opened the pull request:
#### b802bf60e5006c59c715973c2d1041abd33b847c
<pre>
Add GPU acceleration for morphology filters
<a href="https://bugs.webkit.org/show_bug.cgi?id=304758">https://bugs.webkit.org/show_bug.cgi?id=304758</a>
<a href="https://rdar.apple.com/167295476">rdar://167295476</a>

Reviewed by Simon Fraser.

This PR implements Core Image-based GPU acceleration for feMorphology
filters. It uses CIMorphologyRectangleMinimum for erosion and
CIMorphologyRectangleMaximum for dilation.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/coreimage/FEMorphologyCoreImageApplier.h: Added.
* Source/WebCore/platform/graphics/coreimage/FEMorphologyCoreImageApplier.mm: Added.
(WebCore::FEMorphologyCoreImageApplier::FEMorphologyCoreImageApplier):
(WebCore::FEMorphologyCoreImageApplier::supportsCoreImageRendering):
(WebCore::FEMorphologyCoreImageApplier::apply const):
* Source/WebCore/platform/graphics/filters/FEMorphology.cpp:
(WebCore::FEMorphology::supportedFilterRenderingModes const):
(WebCore::FEMorphology::createAcceleratedApplier const):
* Source/WebCore/platform/graphics/filters/FEMorphology.h:

Canonical link: <a href="https://commits.webkit.org/304986@main">https://commits.webkit.org/304986@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f2ed550697a613bafd629fb279a39d12ee00bd70

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/137105 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9465 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48392 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144848 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/90078 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/faa6dd3e-55ab-4e50-a37f-96d0e185c4ab) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138977 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/10168 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/9592 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104847 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/90d98f9c-00a1-4a0a-9144-161b93d0c70d) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/140050 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7475 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122862 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85683 "Found 1 new API test failure: WPE/TestWebKitWebContext:/webkit/WebKitWebContext/uri-scheme (failure)") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/905e4049-1ba1-4e2f-b28e-2adc337838e3) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/7112 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4821 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5438 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116461 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/41028 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147604 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/9148 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41593 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/113201 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/9166 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7689 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113531 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28835 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/7035 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/119123 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/63500 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/9196 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/37177 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8920 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72762 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/9137 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8989 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/coreimage/FEMorphologyCoreImageApplier.mm:
```diff
@@ -0,0 +1,93 @@
+
+bool FEMorphologyCoreImageApplier::supportsCoreImageRendering(const FEMorphology& effect)
+{
+    return effect.radiusX() > 0 || effect.radiusY() > 0;
```
I think we should still claim to support CI rendering, and just be pass-through (if we indeed get here).

smfr commented on Source/WebCore/WebCore.xcodeproj/project.pbxproj:
```diff
@@ -5973,7 +5973,6 @@
 		D20E878D2AF44B5E004579F3 /* SwitchMacUtilities.h in Headers */ = {isa = PBXBuildFile; fileRef = D20E878C2AF44B5E004579F3 /* SwitchMacUtilities.h */; };
 		D22B85932EB4EF6D000D2DC3 /* HTMLSelectedContentElement.h in Headers */ = {isa = PBXBuildFile; fileRef = D22B85912EB4EF6D000D2DC3 /* HTMLSelectedContentElement.h */; settings = {ATTRIBUTES = (Private, ); }; };
 		D245B6DD2B679280009AA00E /* HTTPStatusCodes.h in Headers */ = {isa = PBXBuildFile; fileRef = D245B6D02B6791BA009AA00E /* HTTPStatusCodes.h */; settings = {ATTRIBUTES = (Private, ); }; };
-		D26049272EFD23550067D2DA /* GraphicsLayerKeyframeValueList.cpp in Sources */ = {isa = PBXBuildFile; fileRef = D26049262EFD23550067D2DA /* GraphicsLayerKeyframeValueList.cpp */; };
```
Is this right?

Ahmad-S792 replied:
Yes - I was testing something locally on my end, had same. I think Taher patch is doing it as drive-by fix.

nullhook replied:
odd. xcode housekeeping?

smfr commented on Source/WebCore/platform/graphics/coreimage/FEMorphologyCoreImageApplier.mm:
```diff
@@ -0,0 +1,93 @@
+bool FEMorphologyCoreImageApplier::apply(const Filter& filter, std::span<const Ref<FilterImage>> inputs, FilterImage& result) const
+{
+    ASSERT(inputs.size() == 1);
+    auto& input = inputs[0].get();
```
SaferCPP issue here.

smfr commented on Source/WebCore/platform/graphics/coreimage/FEMorphologyCoreImageApplier.mm:
```diff
@@ -0,0 +1,93 @@
+    ASSERT(inputs.size() == 1);
+    auto& input = inputs[0].get();
+
+    auto inputImage = input.ciImage();
```
`RetainPtr inputImage`

smfr commented on Source/WebCore/platform/graphics/coreimage/FEMorphologyCoreImageApplier.mm:
```diff
@@ -0,0 +1,93 @@
+        ? @"CIMorphologyRectangleMinimum"
+        : @"CIMorphologyRectangleMaximum";
+
+    auto morphologyFilter = [CIFilter filterWithName:filterName];
```
`RetainPtr morphologyFilter`

smfr commented on Source/WebCore/platform/graphics/coreimage/FEMorphologyCoreImageApplier.mm:
```diff
@@ -0,0 +1,93 @@
+    [morphologyFilter setValue:@(kernelWidth) forKey:@"inputWidth"];
+    [morphologyFilter setValue:@(kernelHeight) forKey:@"inputHeight"];
+
+    result.setCIImage(morphologyFilter.outputImage);
```
`[morphologyFilter.outputImage]` to avoid the need to use `.get()`

nullhook committed:
Add GPU acceleration for morphology filters
https://bugs.webkit.org/show_bug.cgi?id=304758
rdar://167295476

Reviewed by Simon Fraser.

This PR implements Core Image-based GPU acceleration for feMorphology
filters. It uses CIMorphologyRectangleMinimum for erosion and
CIMorphologyRectangleMaximum for dilation.

* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/coreimage/FEMorphologyCoreImageApplier.h: Added.
* Source/WebCore/platform/graphics/coreimage/FEMorphologyCoreImageApplier.mm: Added.
(WebCore::FEMorphologyCoreImageApplier::FEMorphologyCoreImageApplier):
(WebCore::FEMorphologyCoreImageApplier::supportsCoreImageRendering):
(WebCore::FEMorphologyCoreImageApplier::apply const):
* Source/WebCore/platform/graphics/filters/FEMorphology.cpp:
(WebCore::FEMorphology::supportedFilterRenderingModes const):
(WebCore::FEMorphology::createAcceleratedApplier const):
* Source/WebCore/platform/graphics/filters/FEMorphology.h:

Canonical link: https://commits.webkit.org/304986@main
